directives/visitor.go

```diff
@@ -0,0 +1,17 @@
+package directives
+
+import (
+	"context"
+
+	"github.com/graph-gophers/graphql-go/types"
+)
+
+// Visitor defines the interface that clients should use to implement a Directive
+// see the graphql.DirectiveVisitors() Schema Option
+type Visitor interface {
+	// Before() is always called when the operation includes a directive matching this implementation's name
+	// Errors in Before() will prevent field resolution
```

Comment:
Let's make sure this is the error behaviour we want before we merge this. This behaviour was here in the earlier version of this PR and I left it unchanged.

**Regular Method Resolvers**

Pros:
1) This lets implementations have some control over whether errors short-circuit resolution. For example, in case of an auth directive, that may be nice–graphql servers may want to abandon un-authorized requests to save CPU cycles.

Cons:
1) May cause surprising results in production. In the case of a directive with a network dependency, fields with directives may fail to resolve during a dependency outage.

It would be nice if we could provide the error status of the `Before()` hook to the resolver and provide the aggregate error state to the `After()` hook. The after hook could accept another argument `err` which it could use to decide whether or not to do its work. For the field resolver case, because we can't just change the API for resolvers, using `ctx` is our best option.

** Struct Resolvers **

I think the behaviour we have right now is the best option. Struct resolvers can't make decisions based on the error state of the directive implementation, so we should bail out for them. We could implement the After() hook for aggregate error information (before hook + field.HasError).

Reply:
Yes, struct fields should be ignored. I also agree with you that the context is how the `Before` and `After` methods can communicate.